### PR TITLE
Proj. 7, Pt. 1: removed compiler errors and added some includes.

### PR DESCRIPTION
**Pull request by hilberturi**
Now there are linker errors.

### CODE REVIEW

File: Character.cpp [hunk 1]
```diff
@@ -98,20 +98,43 @@ void Character::attackInternal(Character& other)
...
+        
+        attackDamage *= 1.1;
+        *initialAttackDamage = attackDamage;
+
+        if (armor < *initialArmorLevel)
+        {
+            armor = *initialArmorLevel;
+        }
+        armor *= 1.1;
+        *initialArmorLevel = armor;
```
Comment (matkatmusic) [inline Character.cpp]: You are doing the same thing three times with different pairs of variables.
Recognize the pattern.
Don't repeat yourself.
Write a free function and call it three times.  The argument list should not include `std::unique_ptr`.
- [x] revise

File: Character.cpp [hunk 2]
```diff
@@ -1,18 +1,20 @@
...
     armor(armor_),
     attackDamage(attackDamage_)
 {
     initialHitPoints.reset( new int(hitPoints) );
     initialArmorLevel.reset( new int( armor) );
     initialAttackDamage.reset( new int( attackDamage) );
+
+    defensiveItems = makeDefensiveItems(numDefensiveItems);
+    helpfulItems = makeHelpfulItems(numHelpfulItems);
 }
```
Comment (matkatmusic) [inline Character.cpp]: Do not change the base class
- [x] revert

File: Character.h [hunk 3]
```diff
@@ -5,11 +5,12 @@
...
 #include <iostream>
 
-#include "Item.h"
+#include "Item.h" // Utility.h prodives Item.h but I've read such hidden includes are bad style
+#include "Utility.h" 
 
 struct Character
 {
-    Character(int hp, int armor_, int attackDamage_ );
+    Character(int hp, int armor_, int attackDamage_, int numDefensiveItems = chooseItemCount(), int numHelpfulItems = chooseItemCount());
```
Comment (matkatmusic) [inline Character.h]: Do not change the base class
- [x] revert

File: DragonSlayer.cpp [hunk 4]
```diff
@@ -1,23 +1,34 @@
...
         //note: items are single-use only, so you need to reset it after use.  
         //look in the Character class for how the other item types are reset after use.
-        while( dragon->getHP() > 0 )
+        while ( dragon->getHP() > 0 )
         {
+            if (auto* attackItemPtr = attackItem.get())
+            {
+                attackItemPtr->use (this);
+                attackItem.reset(); //can only be used once!
+            }
```
Comment (matkatmusic) [inline DragonSlayer.cpp]: Do this before all of the attacking.
You are currently doing this **during** the attacking
- [x] revise

File: Utility.cpp [hunk 5]
```diff
@@ -43,66 +49,82 @@ std::string getCharacterStats(Character* ch)
...
+}
+
+static std::default_random_engine randomEngine {static_cast<unsigned>(std::chrono::system_clock::now().time_since_epoch().count())};
+
+static std::uniform_int_distribution<int> itemDistribution {1, 7};
+
+int chooseItemCount()
+{
+    return itemDistribution(randomEngine);
+}
```
Comment (matkatmusic) [inline Utility.cpp]: Casting has not been covered (it's covered in the next project).
Statics have not been covered
- [x] remove this.

File: Utility.h [hunk 6]
```diff
@@ -4,14 +4,17 @@
 #include <memory>
 #include "Item.h"
 
-
 std::vector<std::unique_ptr<Item>> makeHelpfulItems(int num);
 
 std::vector<std::unique_ptr<Item>> makeDefensiveItems(int num);
 
+std::vector<std::unique_ptr<Item>> makeAttackItems(int num);
+
```
Comment (matkatmusic) [inline Utility.h]: The dragon slayer gets one attack item, not a vector of them. This function is unnecessary.
- [x] remove

**Review by matkatmusic — CHANGES_REQUESTED** [review]

File: Character.cpp [hunk 7]
```diff
@@ -88,6 +85,16 @@ int Character::takeDamage(int damage)
...
 
+void restoreAndBoostStats(int& stat, int* initialLevelPtr)
+{
+    if (stat < *initialLevelPtr)
+    {
+        stat = *initialLevelPtr;
+    }
+    stat *= 1.1;
+    *initialLevelPtr = stat;
```
Comment (matkatmusic) [inline Character.cpp]: You're dereferencing 3 times here.
You don't need to.
The first argument to this function is a reference.
Why can't the 2nd argument to this function also be a reference?
- [x] revise

File: Character.cpp [hunk 8]
```diff
@@ -98,34 +105,16 @@ void Character::attackInternal(Character& other)
...
-        if (armor < *initialArmorLevel)
-        {
-            armor = *initialArmorLevel;
-        }
-        armor *= 1.1;
-        *initialArmorLevel = armor;
-        
+        restoreAndBoostStats (hitPoints, initialHitPoints.get());
+        restoreAndBoostStats (attackDamage, initialAttackDamage.get());
+        restoreAndBoostStats (armor, initialArmorLevel.get());
```
Comment (matkatmusic) [inline Character.cpp]: There is no need to `get()` when you can dereference.
See:
https://en.cppreference.com/w/cpp/memory/unique_ptr/operator*
- [x] revise

**Review by matkatmusic — CHANGES_REQUESTED** [review]

File: Character.cpp (hunk 7 above)
Comment (hilberturi) [reply Character.cpp]: Good point, I have now improved this.

File: Character.cpp (hunk 8 above)
Comment (hilberturi) [reply Character.cpp]: The link seems to be broken but I found the page on cppreference.com anyway.
Using get() is now avoded by dereferencing and passing the reference.

**Review by matkatmusic — APPROVED** [review]

**Pull request merged**